cwms-data-api/src/main/java/cwms/cda/data/dao/location/kind/OutletDao.java

```diff
@@ -0,0 +1,240 @@
+                } else {
+                    throw e;
+                }
+            }
+
+
+            LocationGroupDao locGroupDao = new LocationGroupDao(dsl);
+            List<LocationGroup> groups = locGroupDao.getLocationGroups(officeId, RATING_LOC_GROUP_CATEGORY);
+
+            return mapToOutlet(outletStruct, records, groups);
+        });
+    }
+
+    private String getRatingGroupId(String locationId, List<LocationGroup> groups) {
+        return groups.stream()
+                     .filter(group -> group.getAssignedLocations()
+                                           .stream()
+                                           .anyMatch(loc -> loc.getLocationId().equalsIgnoreCase(locationId)))
+                     .findFirst()
+                     .map(LocationGroup::getId)
+                     .orElse(null);
+    }
+
+    public void storeOutlet(ProjectStructure outlet, String ratingGroup, boolean failIfExists) {
+        connection(dsl, conn -> {
+            PROJECT_STRUCTURE_OBJ_T structure = mapToProjectStructure(outlet);
+            CWMS_OUTLET_PACKAGE.call_STORE_OUTLET(DSL.using(conn).configuration(), structure, ratingGroup,
+                                                  OracleTypeMap.formatBool(failIfExists));
+        });
+    }
+
+    public void deleteOutlet(String locationId, String officeId, DeleteRule deleteRule) {
+        connection(dsl, conn -> CWMS_OUTLET_PACKAGE.call_DELETE_OUTLET(DSL.using(conn).configuration(), locationId,
+                                                                       deleteRule.getRule(), officeId));
+    }
+
+    private Map<String, List<CompoundOutletRecord>> retrieveCompoundOutletsForProject(Configuration config,
+                                                                                      String projectId,
+                                                                                      String officeId) {
+        Map<String, List<CompoundOutletRecord>> recordMap = new HashMap<>();
+        //projectId and officeId are used as a mask in RETRIEVE_COMPOUND_OUTLETS, however this usage expects a specific
+        //id, since retrieveOutlets does not use a mask.
+        STR_TAB_TAB_T tabs = CWMS_OUTLET_PACKAGE.call_RETRIEVE_COMPOUND_OUTLETS(config, projectId, officeId);
+
+        if (!tabs.isEmpty()) {
+            STR_TAB_T tab = tabs.get(0);
+            //This table has a minimum of two columns:  office, and project location
+            //Everything else after that is a compound outlet id.  This requires an additional retrieval to determine
+            //the actual compound outlet records for each outlet.
+            for (int i = 2; i < tab.size(); i++) {
+                String compoundOutletId = tab.get(i);
+                try {
+                    List<CompoundOutletRecord> compoundOutletRecords = retrieveCompoundOutlet(config, compoundOutletId,
```

Comment:
You reviewed the pl/sql API and determined an iteration with subsequent DB retrieve was necessary?

Reply:
It'll take a bit more post processing in the java code, but I think we could reduce the db iterations by using the AV_COMPOUND_OUTLET view.  The view seems fairly simple too.